nmostesting/mocks/Auth.py

```diff
@@ -0,0 +1,535 @@
+
+
+GRANT_TYPES = Enum("grant_types", [
+    "authorization_code", "implicit", "refresh_token", "password", "client_credentials"])
+SCOPES = Enum("scopes", ["connection", "node", "query", "registration", "events", "channelmapping"])
+CODE_CHALLENGE_METHODS = Enum("code_challenge_methods", ["plain", "S256"])
+TOKEN_ENDPOINT_AUTH_METHODS = Enum("token_endpoint_auth_methods", [
+    "none", "client_secret_post", "client_secret_basic", "client_secret_jwt", "private_key_jwt"])
+RESPONSE_TYPES = Enum("response_types", ["code", "token"])
+
+
+class AuthException(Exception):
+    def __init__(self, error, description, httpstatus=HTTPStatus.BAD_REQUEST):
+        self.httpstatus = httpstatus
+        self.error = error
+        self.description = description
+
+
+class AuthServer(object):
+    def __init__(self, auth):
+        self.server = None
+        self.server_thread = None
+        self.auth = auth
+
+    def start(self):
+        """Start Authorization server"""
+        if not self.server_thread:
+            ctx = None
+            # placeholder for the certificate
+            cert_file = "test_data/BCP00301/ca/mock_auth_cert.pem"
+            # placeholder for the private key
+            key_file = "test_data/BCP00301/ca/mock_auth_private_key.pem"
+            # generate RSA key and certificate for the mock secondary Authorization server
+            IS10Utils.make_key_cert_files(cert_file, key_file)
+            if ENABLE_HTTPS:
+                # ssl.create_default_context() provides options that broadly correspond to
+                # the requirements of BCP-003-01
+                ctx = ssl.create_default_context()
+                ctx.load_cert_chain(cert_file, key_file)
+                # additionally disable TLS v1.0 and v1.1
+                ctx.options &= ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1
+                # BCP-003-01 however doesn't require client certificates, so disable those
+                ctx.check_hostname = False
+                ctx.verify_mode = ssl.CERT_NONE
+
+            # mock secondary Authorization server
+            auth_app = Flask(__name__)
+            auth_app.debug = False
+            auth_app.config["AUTH_INSTANCE"] = 1
+            self.auth.private_keys = [key_file]
+            port = self.auth.port
+            auth_app.register_blueprint(AUTH_API)
+
+            self.server = make_server("0.0.0.0", port, auth_app, threaded=True, ssl_context=ctx)
+            self.server_thread = threading.Thread(
+                target=self.server.serve_forever)
+            self.server_thread.daemon = True
+            self.server_thread.start()
+
+    def shutdown(self):
+        """Stop Authorization server"""
+        if self.server_thread:
+            self.server.shutdown()
+            self.server_thread.join()
+            self.server_thread = None
+
+
+class Auth(object):
+    def __init__(self, port_increment, version="v1.0"):
+        self.port = PORT_BASE + 9 + port_increment
```

Comment:
Where does the magic `9` come from? Presumably that's ports reserved for other mocks like the Registry instances?

Hmm, actually, I think the Registry mocks are at `PORT_BASE + 100 + 1` to `PORT_BASE + 100 + 6`... the System API mocks are at `PORT_BASE + 300 + 1` to `PORT_BASE + 300 + 6`. The CRL and OCSP are at `PORT_BASE + 7` and `PORT_BASE + 8` though!

We should probably document this somewhere / put these magic numbers somewhere central? I guess don't worry about it in this PR though...

Reply:
Totally agreed, and the magic 9 is after the OCSP server.